tests/models/test_baseoperatormeta.py

```diff
@@ -0,0 +1,64 @@
+import os
+from typing import Any
+from unittest import TestCase
+from unittest.mock import Mock, patch, MagicMock
+
+import pendulum
+from sqlalchemy.orm import Session
+
+from airflow import AirflowException, DAG
+from airflow.models import TaskInstance, DagRun
+from airflow.models.baseoperator import BaseOperator
+from airflow.utils.context import Context
+from airflow.utils.state import DagRunState
+
+
+class HelloWorldOperator(BaseOperator):
+    called = False
+
+    def execute(self, context: Context) -> Any:
+        HelloWorldOperator.called = True
+        return f"Hello {self.owner}!"
+
+
+class IterableSession(Session):
+    def __next__(self):
+        pass
+
+
+class BaseOperatorMetaTestCase(TestCase):
+
+    @patch.dict(os.environ, {'AIRFLOW__CORE__UNIT_TEST_MODE': 'False'})
+    def test_executor_safeguard_when_unauthorized(self):
+        with self.assertRaises(AirflowException):
+            dag = DAG(dag_id="hello_world")
+            context = MagicMock(spec=Context)
+
+            HelloWorldOperator(task_id="task_id", dag=dag).execute(context=context)
+
+    @patch("sqlalchemy.orm.Session.__init__")
+    @patch.dict(os.environ, {'AIRFLOW__CORE__UNIT_TEST_MODE': 'False'})
+    def test_executor_safeguard_when_authorized(self, mock_session: MagicMock):
+        session = MagicMock(spec=IterableSession)
+        mock_session.return_value = session
+        session.__iter__.return_value = iter({})
+
+        dag = DAG(dag_id="hello_world")
+        context = {"params": {}}
```

Comment:
I'm not sure what you tried to achieve here but seems like it might be replaced by [`dag_maker`](https://github.com/apache/airflow/blob/0232ad031890c5f2b1bdbdcd3e28b4c6e16abfb3/tests/conftest.py#L612) fixture.

Reply:
Tried to use the dag_maker fixture, but the issue there is that you need a db while I've made my tests run without the need of a database

Reply:
To ellaborat a bit more, I don't bother about the interactions with the database here, unless we would also need to provide an integration test.  When using the dag_maker fixer, you're bound to use a database (sqlite), which I didn't want to make the test as light (and fast) as possible a I only want to test my decorator, not all the other stuff to make the test runnable.

Reply:
approx 40% of all tests required/use DB connection, one more or one less it is not a big deal especially if it helps avoid to do the same operation in different tests, e.g. create DagRun, task execution or rendering templated fields

This is statistics from the Non-DB tests
```console
10533 passed, 8552 skipped,
```

Reply:
> you're bound to use a database (sqlite)

Small clarification it will use any supported database: Postgres, SQLite, MySQL.
For DB required tests also required specific marker: `@pytest.mark.db`

If you worried about that some tests required DB when in unnecessary, there are hundreds of providers tests where DB is not required and which potentially could be fixed

Reply:
I think I need to add another test case, namely a test case where a PythonOperator is being executed through the TaskInstance but has a callable which manually calls the execute of the HelloWorldOperator, as this is what will happen in real on Airflow and that should raise an AirflowException.

Reply:
> > you're bound to use a database (sqlite)
> 
> Small clarification it will use any supported database: Postgres, SQLite, MySQL. For DB required tests also required specific marker: `@pytest.mark.db`
> 
> If you worried about that some tests required DB when in unnecessary, there are hundreds of providers tests where DB is not required and which potentially could be fixed

Indeed, it's just convienient to me as I can just directly run the test without the need to setup any db and have sqlalchemy do all it's stuff on the ORM, that's why I did it like this.

Reply:
I've refactored the test a bit, but still I wanted it not to use the DB.  I've also read through [this unit_tests documentation ](https://github.com/apache/airflow/blob/main/contributing-docs/testing/unit_tests.rst#best-practices-for-db-tests) and if it's possible to avoid the DB it's encouraged hence why I also want to keep it that way.

I also had as already was suggested need to add a test_mode flag to the decorator to make it testable, changing the unit_test_mode jsut before executing the task was not enough as that would still required some DB interaction.  Now it fully works in an isolated manner as a unit test.

Reply:
Yeah. Having non-db tests is likely better. We shoud avoid db-tests if necessary

Reply:
Indeed but in thise case @Taragolis was right as this cannot be tested as a "simple" unit test as there was to much to mock which would make it unmaintanable (e.g. easy to break) in the future when refactorings would be done, hence I switched to the dag_maker fixture as a db test which makes it much cleaner, concise and easier to read (and maintain).